polkadot/runtime/common/src/paras_registrar/mod.rs

```diff
@@ -404,16 +441,45 @@ pub mod pallet {
 		///
 		/// Can be called by Root, the parachain, or the parachain manager if the parachain is
 		/// unlocked.
+		///
+		/// In case the call is made by the parachain manager or the parachain itself the upgrade
+		/// fee will be charged.
+		///
+		/// The caller will receive a refund or be required to reserve an additional deposit,
+		/// depending on the size of the new validation code.
 		#[pallet::call_index(7)]
 		#[pallet::weight(<T as Config>::WeightInfo::schedule_code_upgrade(new_code.0.len() as u32))]
 		pub fn schedule_code_upgrade(
 			origin: OriginFor<T>,
 			para: ParaId,
 			new_code: ValidationCode,
 		) -> DispatchResult {
-			Self::ensure_root_para_or_owner(origin, para)?;
-			runtime_parachains::schedule_code_upgrade::<T>(para, new_code, SetGoAhead::No)?;
-			Ok(())
+			Self::ensure_root_para_or_owner(origin.clone(), para)?;
+
+			let fee_payer = if ensure_root(origin.clone()).is_ok() {
+				// Root doesn't pay. This, also means that system parachains do not pay for upgrade
+				// fees.
+				None
+			} else if LegacyParas::<T>::get(para).is_some() {
+				// Each legacy para is permitted to perform one upgrade for free.
+				//
+				// This is introduced to  avoid causing a breaking change to the system once para
+				// upgrade fees are required.
```

Comment:
By only allowing one upgrade we are not fulfilling that requirement. What if a legacy para does two updates?